File: siden-core/src/main/java/ninja/siden/internal/RendererSelector.java
/*
 * Copyright 2014 SATO taichi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package ninja.siden.internal;

import io.undertow.io.IoCallback;
import io.undertow.server.HttpServerExchange;
import io.undertow.server.handlers.resource.URLResource;
import io.undertow.util.MimeMappings;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.Reader;
import java.net.URI;
import java.net.URL;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.function.Predicate;

import ninja.siden.Config;
import ninja.siden.Renderer;

import org.xnio.IoUtils;
import org.xnio.OptionMap;
import org.xnio.streams.ReaderInputStream;
import org.xnio.streams.Streams;

/**
 * @author taichi
 */
public class RendererSelector<T> implements Renderer<T> {

	List<SelectableRenderer<T>> renderers;

	public RendererSelector() {
		this(defaultRenderers());
	}

	public RendererSelector(List<SelectableRenderer<T>> renderers) {
		this.renderers = renderers;
	}

	public static <T> List<SelectableRenderer<T>> defaultRenderers() {
		return Arrays.asList(new StringRenderer<T>(), new FileRenderer<T>(),
				new PathRenderer<T>(), new FileChannelRenderer<T>(),
				new ByteArrayRenderer<T>(), new ByteBufferRenderer<T>(),
				new URIRenderer<T>(), new URLRenderer<T>(),
				new ReaderRenderer<T>(), new InputStreamRenderer<T>(),
				new CharSequenceRenderer<T>(), new ToStringRenderer<T>());
	}

	@Override
	public void render(T model, HttpServerExchange sink) throws IOException {
		for (SelectableRenderer<T> pr : renderers) {
			if (pr.test(model)) {
				pr.render(model, sink);
				return;
			}
		}
	}

	interface SelectableRenderer<T> extends Renderer<T>, Predicate<T> {
	}

	static class ReaderRenderer<T> implements SelectableRenderer<T> {
		final InputStreamRenderer<T> delegate = new InputStreamRenderer<T>();

		@Override
		public boolean test(Object model) {
			return Reader.class.isAssignableFrom(model.getClass());
		}

		@Override
		public void render(Object model, HttpServerExchange sink)
				throws IOException {
			Reader reader = (Reader) model;
			SecurityHandler.addContentType(sink);
			OptionMap config = sink.getAttachment(Core.CONFIG);
			delegate.render(
					new ReaderInputStream(reader, config.get(Config.CHARSET)),
					sink);
		}
	}

	static class InputStreamRenderer<T> implements SelectableRenderer<T> {

		@Override
		public boolean test(Object model) {
			return InputStream.class.isAssignableFrom(model.getClass());
		}

		@Override
		public void render(Object model, HttpServerExchange sink)
				throws IOException {
			SecurityHandler.addContentType(sink);
			Renderer.ofStream((Object o, OutputStream out) -> {
				InputStream in = (InputStream) o;
				try {
					Streams.copyStream(in, out, false);
				} finally {
					IoUtils.safeClose(in);
				}
			}).render(model, sink);
		}
	}

	static class URIRenderer<T> implements SelectableRenderer<T> {
		final Renderer<URL> delegate = new URLRenderer<URL>();

		@Override
		public boolean test(Object t) {
			return URI.class == t.getClass();
		}

		@Override
		public void render(Object model, HttpServerExchange sink)
				throws IOException {
			URI uri = (URI) model;
			this.delegate.render(uri.toURL(), sink);
		}
	}

	static class URLRenderer<T> implements SelectableRenderer<T> {

		@Override
		public boolean test(Object t) {
			return URL.class == t.getClass();
		}

		@Override
		public void render(Object model, HttpServerExchange sink)
				throws IOException {
			URL url = (URL) model;
			OptionMap config = sink.getAttachment(Core.CONFIG);
			MimeMappings mm = config.get(Config.MIME_MAPPINGS);
			// TODO proxy?
			URLResource resource = new URLResource(url, url.openConnection(),
					url.getPath());
			SecurityHandler.addContentType(sink, resource.getContentType(mm));
			resource.serve(sink.getResponseSender(), sink,
					IoCallback.END_EXCHANGE);
		}
	}

	static class PathRenderer<T> implements SelectableRenderer<T> {
		Renderer<URL> delegate = new URLRenderer<URL>();

		@Override
		public boolean test(Object t) {
			return Path.class.isAssignableFrom(t.getClass());
		}

		@Override
		public void render(Object model, HttpServerExchange sink)
				throws IOException {
			Path path = (Path) model;
			this.delegate.render(path.toUri().toURL(), sink);
		}
	}

	static class FileRenderer<T> implements SelectableRenderer<T> {
		Renderer<URL> delegate = new URLRenderer<URL>();

		@Override
		public boolean test(Object t) {
			return File.class.isAssignableFrom(t.getClass());
		}

		@Override
		public void render(Object model, HttpServerExchange sink)
				throws IOException {
			File file = (File) model;
			this.delegate.render(file.toURI().toURL(), sink);
		}
	}

	static class FileChannelRenderer<T> implements SelectableRenderer<T> {

		@Override
		public boolean test(Object t) {
			return FileChannel.class.isAssignableFrom(t.getClass());
		}

		@Override
		public void render(Object model, HttpServerExchange sink)
				throws IOException {
			FileChannel channel = (FileChannel) model;
			SecurityHandler.addContentType(sink);
			sink.getResponseSender().transferFrom(channel,
					IoCallback.END_EXCHANGE);
		}
	}

	static class ByteBufferRenderer<T> implements SelectableRenderer<T> {

		@Override
		public boolean test(Object t) {
			return ByteBuffer.class.isAssignableFrom(t.getClass());
		}

		@Override
		public void render(Object model, HttpServerExchange sink)
				throws IOException {
			ByteBuffer s = (ByteBuffer) model;
			SecurityHandler.addContentType(sink);
			sink.getResponseSender().send(s);
		}
	}

	static class ByteArrayRenderer<T> implements SelectableRenderer<T> {

		@Override
		public boolean test(Object t) {
			return byte[].class == t.getClass();
		}

		@Override
		public void render(Object model, HttpServerExchange sink)
				throws IOException {
			byte[] ary = (byte[]) model;
			SecurityHandler.addContentType(sink);
			sink.getResponseSender().send(ByteBuffer.wrap(ary));
		}
	}

	public static class StringRenderer<T> implements SelectableRenderer<T> {

		@Override
		public boolean test(Object t) {
			return String.class == t.getClass();
		}

		@Override
		public void render(Object model, HttpServerExchange sink)
				throws IOException {
			OptionMap config = sink.getAttachment(Core.CONFIG);
			String s = model.toString();
			SecurityHandler.addContentType(
					sink,
					String.format("text/plain; charset=%s",
							config.get(Config.CHARSET)));
			sink.getResponseSender().send(s, config.get(Config.CHARSET));
		}
	}

	static class CharSequenceRenderer<T> implements SelectableRenderer<T> {
		StringRenderer<T> deleagte = new StringRenderer<T>();

		@Override
		public boolean test(Object t) {
			return CharSequence.class.isAssignableFrom(t.getClass());
		}

		@Override
		public void render(Object model, HttpServerExchange sink)
				throws IOException {
			this.deleagte.render(model, sink);
		}
	}

	static class ToStringRenderer<T> implements SelectableRenderer<T> {
		StringRenderer<T> deleagte = new StringRenderer<T>();

		@Override
		public boolean test(Object t) {
			return true;
		}

		@Override
		public void render(Object model, HttpServerExchange sink)
				throws IOException {
			this.deleagte.render(Objects.toString(model), sink);
		}
	}
}

File: siden-core/src/main/kotlin/ninja/siden/internal/RendererSelector.kt
/*
 * Copyright 2014 SATO taichi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package ninja.siden.internal

import io.undertow.io.IoCallback
import io.undertow.server.HttpServerExchange
import io.undertow.server.handlers.resource.URLResource
import ninja.siden.Config
import ninja.siden.Renderer
import org.jboss.logging.Logger
import org.xnio.IoUtils
import org.xnio.streams.ReaderInputStream
import org.xnio.streams.Streams
import java.io.File
import java.io.InputStream
import java.io.OutputStream
import java.io.Reader
import java.net.URI
import java.net.URL
import java.nio.ByteBuffer
import java.nio.channels.FileChannel
import java.nio.file.Path
import java.util.*
import java.util.function.Predicate

/**
 * @author taichi
 */
class RendererSelector<T> @JvmOverloads constructor(
        val renderers: List<RendererSelector.SelectableRenderer> = RendererSelector.defaultRenderers()) : Renderer<T> {

    override fun render(model: T, sink: HttpServerExchange) {
        for (pr in renderers) {
            if (pr.test(model)) {
                pr.render(model as Any, sink)
                return
            }
        }
    }

    interface SelectableRenderer : Renderer<Any>, Predicate<Any>

    internal class ReaderRenderer : SelectableRenderer {
        val delegate = InputStreamRenderer()

        override fun test(model: Any?) = model is Reader

        override fun render(model: Any, sink: HttpServerExchange) {
            val reader = model as Reader
            SecurityHandler.addContentType(sink)
            val config = sink.getAttachment(Core.CONFIG)
            delegate.render(ReaderInputStream(reader, config.get(Config.CHARSET)), sink)
        }
    }

    internal class InputStreamRenderer : SelectableRenderer {
        override fun test(model: Any?) = model is InputStream

        override fun render(model: Any, sink: HttpServerExchange) {
            SecurityHandler.addContentType(sink)
            ninja.siden.Renderer.ofStream { o: Any, out: OutputStream ->
                val ins = o as InputStream
                try {
                    Streams.copyStream(ins, out, false)
                } finally {
                    IoUtils.safeClose(ins)
                }
            }.render(model, sink)
        }
    }

    internal class URIRenderer : SelectableRenderer {
        val delegate = URLRenderer()

        override fun test(t: Any?) = t is URI

        override fun render(model: Any, sink: HttpServerExchange) {
            val uri = model as URI
            this.delegate.render(uri.toURL(), sink)
        }
    }

    internal class URLRenderer : SelectableRenderer {

        override fun test(t: Any?) = t is URL

        override fun render(model: Any, sink: HttpServerExchange) {
            val url = model as URL
            val config = sink.getAttachment(Core.CONFIG)
            val mm = config.get(Config.MIME_MAPPINGS)
            // TODO proxy?
            val resource = URLResource(url, url.openConnection(), url.path)
            SecurityHandler.addContentType(sink, resource.getContentType(mm))
            resource.serve(sink.responseSender, sink, IoCallback.END_EXCHANGE)
        }
    }

    internal class PathRenderer : SelectableRenderer {
        var delegate = URLRenderer()

        override fun test(t: Any?) = t is Path

        override fun render(model: Any, sink: HttpServerExchange) {
            val path = model as Path
            this.delegate.render(path.toUri().toURL(), sink)
        }
    }

    internal class FileRenderer : SelectableRenderer {
        var delegate = URLRenderer()

        override fun test(t: Any?) = t is File

        override fun render(model: Any, sink: HttpServerExchange) {
            val file = model as File
            if (file.exists()) {
                this.delegate.render(file.toURI().toURL(), sink)
            } else {
                LOG.error(file.absolutePath + " not exists")
                sink.responseCode = 500
            }
        }
    }

    internal class FileChannelRenderer : SelectableRenderer {

        override fun test(t: Any?) = t is FileChannel

        override fun render(model: Any, sink: HttpServerExchange) {
            val channel = model as FileChannel
            SecurityHandler.addContentType(sink)
            sink.responseSender.transferFrom(channel, IoCallback.END_EXCHANGE) // TODO close
        }
    }

    internal class ByteBufferRenderer : SelectableRenderer {

        override fun test(t: Any?) = t is ByteBuffer

        override fun render(model: Any, sink: HttpServerExchange) {
            val s = model as ByteBuffer
            SecurityHandler.addContentType(sink)
            sink.responseSender.send(s)
        }
    }

    internal class ByteArrayRenderer : SelectableRenderer {

        override fun test(t: Any?) = t is ByteArray

        override fun render(model: Any, sink: HttpServerExchange) {
            val ary = model as ByteArray
            SecurityHandler.addContentType(sink)
            sink.responseSender.send(ByteBuffer.wrap(ary))
        }
    }

    class StringRenderer : SelectableRenderer {

        override fun test(t: Any?) = t is String

        override fun render(model: Any, sink: HttpServerExchange) {
            val config = sink.getAttachment(Core.CONFIG)
            val s = model.toString()
            SecurityHandler.addContentType(sink,
                    config.get(Config.DEFAULT_CONTENT_TYPE))
            sink.responseSender.send(s, config.get(Config.CHARSET))
        }
    }

    internal class CharSequenceRenderer : SelectableRenderer {
        var deleagte = StringRenderer()

        override fun test(t: Any?) = t is CharSequence

        override fun render(model: Any, sink: HttpServerExchange) {
            this.deleagte.render(model, sink)
        }
    }

    internal class ToStringRenderer : SelectableRenderer {
        var deleagte = StringRenderer()

        override fun test(t: Any) = true

        override fun render(model: Any, sink: HttpServerExchange) {
            this.deleagte.render(Objects.toString(model), sink)
        }
    }

    companion object {

        internal val LOG = Logger.getLogger(RendererSelector::class.java)

        internal fun defaultRenderers(): List<SelectableRenderer> {
            return Arrays.asList(StringRenderer(), FileRenderer(),
                    PathRenderer(), FileChannelRenderer(),
                    ByteArrayRenderer(), ByteBufferRenderer(),
                    URIRenderer(), URLRenderer(),
                    ReaderRenderer(), InputStreamRenderer(),
                    CharSequenceRenderer(), ToStringRenderer())
        }
    }
}
